services/libs/aster-std/src/syscall/eventfd.rs

```diff
@@ -0,0 +1,239 @@
+//! flag is set.). The read/write operations may be blocked based on certain requirements.
+//!
+//! For more detailed information about this syscall, refer to the man 2 eventfd documentation.
+
+use super::{SyscallReturn, SYS_EVENTFD, SYS_EVENTFD2};
+use crate::events::{IoEvents, Observer};
+use crate::fs::file_handle::FileLike;
+use crate::fs::utils::{CreationFlags, StatusFlags};
+use crate::log_syscall_entry;
+use crate::prelude::*;
+use crate::process::signal::{Pauser, Pollee, Poller};
+
+pub fn sys_eventfd(init_val: u64) -> Result<SyscallReturn> {
+    log_syscall_entry!(SYS_EVENTFD);
+    debug!("init_val = 0x{:x}", init_val);
+
+    let event_file = {
+        let flags = Flags::empty();
+        EventFile::new(init_val, flags)
+    };
+
+    let fd = {
+        let current = current!();
+        let mut file_table = current.file_table().lock();
+        // TODO: deal with close_on_exec
+        file_table.insert(Arc::new(event_file))
+    };
+
+    Ok(SyscallReturn::Return(fd as _))
+}
+
+pub fn sys_eventfd2(init_val: u64, flags: u32) -> Result<SyscallReturn> {
+    log_syscall_entry!(SYS_EVENTFD2);
+    trace!("raw flags = {}", flags);
+    let flags = Flags::from_bits(flags)
+        .ok_or_else(|| Error::with_message(Errno::EINVAL, "unknown flags"))?;
+    debug!("init_val = 0x{:x}, flags = {:?}", init_val, flags);
+
+    let event_file = EventFile::new(init_val, flags);
+    let fd = {
+        let current = current!();
+        let mut file_table = current.file_table().lock();
+        // TODO: deal with close_on_exec
+        file_table.insert(Arc::new(event_file))
+    };
+
+    Ok(SyscallReturn::Return(fd as _))
+}
+
+bitflags! {
+    struct Flags: u32 {
+        const EFD_SEMAPHORE = 1;
+        const EFD_CLOEXEC = CreationFlags::O_CLOEXEC.bits();
+        const EFD_NONBLOCK = StatusFlags::O_NONBLOCK.bits();
+    }
+}
+
+struct EventFile {
+    counter: Mutex<u64>,
+    pollee: Pollee,
+    flags: Mutex<Flags>,
+    writing_pauser: Arc<Pauser>,
+}
+
+impl EventFile {
+    const MAX_COUNTER_VALUE: u64 = u64::MAX - 1;
+
+    fn new(init_val: u64, flags: Flags) -> Self {
+        let counter = Mutex::new(init_val);
+        let pollee = Pollee::new(IoEvents::OUT);
+        let writing_pauser = Pauser::new();
+        Self {
+            counter,
+            pollee,
+            flags: Mutex::new(flags),
+            writing_pauser,
+        }
+    }
+
+    fn is_nonblocking(&self) -> bool {
+        self.flags.lock().contains(Flags::EFD_NONBLOCK)
+    }
+
+    fn update_io_state(&self, counter: &MutexGuard<u64>) {
+        if **counter == 0 {
+            self.pollee.del_events(IoEvents::IN);
+        } else {
+            self.pollee.add_events(IoEvents::IN);
+        }
+
+        // if it is possible to write a value of at least "1" without
+        // blocking, the file is writable
+        if **counter < Self::MAX_COUNTER_VALUE {
+            self.pollee.add_events(IoEvents::OUT);
```

Comment:
For efficiency, combine the two calls to `add_events` into one. The same argument applies to `del_events`.